### DISCUSSION TITLE
How to force BN to render freshly-defined variables from python

### DISCUSSION THREAD

**Discussion by mostobriv** [Vector35/binaryninja-api#3414] [Q&A] [answered]
Hi! Migrating my IDA gui scripts wrote simple plugin to implement string creation from selection, but found strange behavior of BN when defining global variable from python: 
1 - BN don't render string literal until cursor is moved
2 - sometimes BN just don't render newly created strings at all until view is changed, also selection of raw-hex in linear view works incorrectly (selection doesn't match position of cursor)

```python
from binaryninjaui import (UIAction, UIActionHandler)

def handle_action(context):
    current_selection_range = context.view.getSelectionOffsets()
    bv = context.binaryView
    t = bv.parse_type_string(f"char[{current_selection_range[1] - current_selection_range[0]}]")
    bv.define_user_data_var(current_selection_range[0], t[0])


UIAction.registerAction("Enhanced make string")
UIActionHandler.globalActions().bindAction("Enhanced make string", UIAction(handle_action))
```

actually same behavior if define string from python-console

So the question is how can i refresh current view from python to force BN render created variable?

**Comment by psifertex** [comment, marked as answer]
RE: 2) can you screenshot what you mean when you say the selection doesn't match? Do you mean the offset and bytes in the bottom-right don't match what you actually have selected? 

RE: 1) you may need to trigger [update_analysis_and_wait](https://api.binary.ninja/binaryninja.binaryview-module.html?highlight=update_analysis#binaryninja.binaryview.BinaryView.update_analysis_and_wait). The scripting console and snippets do this for you automatically but plugins do not since it's not necessarily obvious that all scripts should immediately trigger analysis so it's not forced on any plugin. However, there is a known visual issue where sometimes after making changes you have to navigate the UI to get it freshed and we don't have a known fix for it at this time beyond just moving or hitting space twice to make the view refresh.

**Reply by mostobriv** [reply]
About "strange" selection, here is quick demo, because idk how to show it just with screenshot: https://imgur.com/a/YTDhS6p

but guess this is not relevant as it also can be fixed when variable appears visually

**Reply by psifertex** [reply]
I'm afraid I didn't see anything strange even after watching it several times in slow-motion. The addresses appear to be correct at all times based on what was selected. 

Can you be more specific as to what you expected to happen that didn't, or what happened that you didn't expect to happen?
